gamemodes/terrortown/gamemode/shared/sh_equip_items.lua

```diff
@@ -876,43 +870,105 @@ local function ValueToKey(tbl)
 end
 
 ---
--- Cleans up the structure of the CanBuy table of all weapons and items.
+-- Cleans up the structure of the CanBuy table of the given equipment.
 -- After calling this, all keys will be equal to their value. This allows access in O(1) rather than O(n).
 -- table.HasValue is still supported, but please just check if the key is not nil instead.
+-- @param table eq Equipment you clean the buy indices from
 -- @internal
 -- @realm shared
-local function CleanUpDefaultCanBuyIndices()
-	local itms = items.GetList()
+-- @local
+local function CleanUpDefaultCanBuyIndices(eq)
+	eq.CanBuy = eq.CanBuy or {}
 
-	-- load items
-	for i = 1, #itms do
-		local itm = itms[i]
+	ValueToKey(eq.CanBuy)
+end
 
-		itm.CanBuy = itm.CanBuy or {}
+---
+-- Initializes the default equipment for traitors and detectives
+-- @param table eq Equipment you initialize
+-- @internal
+-- @realm shared
+function InitDefaultEquipment(eq)
+	CleanUpDefaultCanBuyIndices(eq)
+	InitDefaultEquipmentForRole(TRAITOR, eq)
+	InitDefaultEquipmentForRole(DETECTIVE, eq)
+end
 
-		ValueToKey(itm.CanBuy)
-	end
+---
+-- Resets the default equipment for a role
+-- @param roleData roleData Role you reset for
+-- @param table eq Equipment you reset
+-- @internal
+-- @realm shared
+local function ResetDefaultEquipmentForRole(roleData, eq)
+	local tbl = roleData.fallbackTable or {}
+	local tblSize = #tbl
+
+	-- is a buyable equipment to load info from
+	if not eq or eq.Doublicated then return end
+
+	local base = GetEquipmentBase(eq)
+	if not base then return end
+
+	local addEquipment = tobool(eq.CanBuy and eq.CanBuy[roleData.index])
 
-	local sweps = weapons.GetList()
+	if addEquipment then
+		local tableHasValue = false
 
-	-- load sweps
-	for i = 1, #sweps do
-		local wep = sweps[i]
+		for i = 1, tblSize do
+			if tbl[i].id == base.id then
+				tableHasValue = true
 
-		wep.CanBuy = wep.CanBuy or {}
+				break
+			end
+		end
+
+		if not tableHasValue then
+			tbl[tblSize + 1] = base
+		end
+	else
+		local counter = 0
+
+		for i = 1, tblSize do
+			-- Skip the equipment that should be removed
+			if tbl[i].id == base.id then continue end
+
+			counter = counter + 1
+
+			if counter == i then continue end
+
+			-- Replace skipped equipments with higher indices
+			tbl[counter] = tbl[i]
+		end
 
-		ValueToKey(wep.CanBuy)
+		local diff = tblSize - counter
+
+		-- Remove last table entries
+		if diff > 0 then
+			for i = 0, diff - 1 do
+				table.remove(tbl, tblSize - i)
+			end
+		end
```

Comment:
Okay, added some documentation, hope this helps to understand. And added a second argument to table.remove to make it clear, that the last item is removed.